### ISSUE TITLE
Allow color settings to be seperated into multiple fieldsets

### ISSUE THREAD

**Issue by hosef** [backdrop/backdrop-issues#3257]
## Describe your issue or idea
Currently, the color module puts all of the color settings into a single fieldset and has no way to change that. I think it would provide a lot of options for theme creators if we allowed them to group relevant color settings.
Currently, the theme must create a theme-settings.php file to have a color form at all, so we could just allow theme developers to create empty arrays for the color module to place it's elements into. the color module also uses the HTML IDs to find the elements, which would break the color module if you tried to move the elements or didn't name the fieldsets correctly, so we can switch to using data attributes to make it more flexible.

---
PR by @hosef: https://github.com/backdrop/backdrop/pull/2409

**Comment by quicksketch** [comment]
I've read over the PR at https://github.com/backdrop/backdrop/pull/2285 and I think it's a mixed bag. The use of `data-` attributes over IDs is definitely a plus, but I don't like the new recursive locating of colors, or how the form array is flattened out.

I think the primary concern here (being able to put color fields in different fieldsets) could be overcome with the current code by theme's specifying `#parents`.

For example, both of these forms will result in the same `$form_state['values']`:

```
$form['fieldset']['foo'] = array(
  '#type' => 'textfield',
);
$form['fieldset']['bar'] = array(
  '#type' => 'textfield',
);
```

Or this where fields are separated:
```
$form['fieldset1']['foo'] = array(
  '#type' => 'textfield',
  '#parents' => array('fieldset', 'foo'),
);
$form['fieldset2']['bar'] = array(
  '#type' => 'textfield',
  '#parents' => array('fieldset', 'bar'),
);
```

Both of these will end up with `$form_state['values']` as:
```
array(
  'fieldset' => array(
    'foo' => 'example1',
    'bar' => 'example2',
  ),
);
```

Setting `#parents` like this actually also affects the generated HTML `id` attribute, so I think that we actually already have the ability requested here, fields can already be separated into different fieldsets while maintaining the exact same IDs and $form_state values as we have currently. Including the `data-` attributes changes would still be good though.

**Comment by hosef** [comment]
I created a new pull request that removes the recursive logic and just sets `#parents` so that the submit handler does not need to be changed.

The new way of specifying where the form elements are is to call a new function that I added that also stores a list of which color elements were already placed so that the Color module does not try to place them twice.

**Comment by quicksketch** [comment]
The static variable for holding rendered elements seems strange to me. The logic is hard to follow as to how that variable gets set and what `_color_get_color_element()` returns. At the very least, we'll need to make `_color_get_color_element()` a public function (remove the leading underscore) so that its clear that it can and should be called by themes like Basis.

**Comment by hosef** [comment]
The static variable doesn't hold rendered elements, it only stores a list of the element names that were already used. This happens so that you can move some elements into a separate field set, a leave the old behavior for some of them.
I am doing that because I don't have any other ideas about how the theme can let the color module know what it is doing and also maintain backwards compatibility.

I changed the function name as you said, and I added some more comments.

**Comment by hosef** [comment]
I updated `color_get_color_element()` to use a key in the form array to store the list of which elements have been generated already instead of sing a static variable.

**Comment by jenlampton** [comment]
I tested in the sandbox and this is working for me as expected. There's one failing test in PHP7 `Comment: Comment interface`, but I'm not sure how that's related. 

I left a few notes on the PR, but it looks like there are a few blank lines being added where they are not needed. Other than those and the failing test, this looks good to me!

**Comment by klonos** [comment]
@jenlampton I have been tormented by this `comment-new class not found` test fail for so long 😞😠 ...tis random.

**Comment by klonos** [comment]
While at it, should we also fix the spacing between the label and the drop-down in the `COLOR SCHEME` fieldset please?

**Comment by hosef** [comment]
@klonos Is there an issue for the comment tests? I get an error in there sometimes on my local, but it has a different message.

**Comment by hosef** [comment]
@jenlampton Are the differences from Drupal coding standards documented somewhere? I found https://api.backdropcms.org/php-standards which doesn't mention new lines, but it points to https://www.drupal.org/docs/develop/standards which says that PHP open tags at the top of the files should always be followed by a blank line.

**Comment by klonos** [comment]
@hosef please have a look here: https://github.com/backdrop/backdrop-issues/issues/1478

**Comment by hosef** [comment]
I removed the extra blank lines and refreshed the PR.

For some reason ZenCI can't find the test script, but I think that is unrelated, so I am going to mark this Needs Review again.

**Comment by herbdool** [comment]
Tested and this looks good to me.

**Comment by quicksketch** [comment]
To help facilitate other color-related issues like https://github.com/backdrop/backdrop/pull/2299, I've merged this into 1.12.x as well as 1.x. Changes seem fairly limited now, just a more flexible implementation.

Thanks @hosef, @herbdool, and @jenlampton for your collaboration on this issue!
